### ISSUE TITLE
SCA

### ISSUE THREAD

**Issue by jasoncharnes** [pay-rails/pay#84]
Now that things are slowing down at work from implementing [Strong Customer Authentication (SCA)](https://stripe.com/docs/strong-customer-authentication), I want to start the process of trying to add it to Pay.

While I'm familiar with the regular _and_ how Stripe implements it, I'm not sure (yet) how Braintree handles SCA. With that said, my contribution (at first) will likely be a Stripe-focused solution. If anyone would like to rubber-duck to make sure the approach with Stripe makes sense at a high-level since we also implement SCA for Braintree, too. 👍 

Currently, pre-SCA (the good ol' days), you would receive a `token` from Stripe, send it to a `Stripe::Charge`, `Stripe::Customer`, or `Stripe::Subscription` (right?). 

I just spent a massive amount of time regurgitating the Stripe docs in this issue, which is irrelevant to the PR (because I link to it above). I deleted it and what I would like to do, instead, is talk about the potential API within Pay.

Before I get started, I've considered making SCA an optional buy-in, at first. I don't have all the logistics worked out in my head, yet.

```ruby
Pay.setup do |config|
  config.use_strong_customer_authentication = false
end
```

## Creating a charge with Stripe

I think we can keep the [existing API](https://github.com/jasoncharnes/pay#creating-a-charge). Though, I think it might be worth making `payment_method_id` and `payment_intent_id` arguments to `charge` versus attributes on the object.

```ruby
user.charge(5000, payment_method: "pm_card_visa")
```

But, instead of returning the charge*, return a `Pay::Payment` object (since `Pay::Charge` is a model with a table and all that) that can respond to SCA-related methods.
###### * unless they choose not to opt-in to SCA

```ruby
# Does not require authentication
charge = user.charge(5000, payment_method: "pm_card_visa")
charge.success? #=> true, no authentication required

# Requires authentication
charge = user.charge(5000, payment_method: "pm_card_visa")
charge.success? #=> false, requires authentication
charge.requires_action? #=> true (user can take this and do something with it)
charge.errors #=> ['This payment requires authentication'] 🤷‍♂ 
charge.payment_intent_id #=> "pi_id"
charge.client_secret #=> "client_secret"
# ... go to the client and authenticate
charge = user.charge(5000, payment_intent: "pi_id")
charge.success? #=> true
```

## Creating a subscription with Stripe

I'd like to (mostly) keep the [existing API](https://github.com/jasoncharnes/pay#creating-a-subscription).

```ruby
user.subscribe(payment_method: "pm_card_visa")
```

Here, we continue to return `Pay::Subscription` and add the appropriate methods.

```ruby
# Does not require authentication
subscription = user.subscribe(payment_method: "pm_card_visa")
subscription.active? #=> true, no authentication required

# Requires authentication
subscription = user.subscribe(payment_method: "pm_card_visa")
subscription.active? #=> false, requires authentication
subscription.requires_action? #=> true (user can take this and do something with it)
subscription.errors #=> ['This subscription requires authentication'] 🤷‍♂ 
subscription.payment_intent_id #=> "pi_id"
subscription.payment_intent_client_secret #=> "client_secret"
# ... go to the client and authenticate, but have a webhook that handles making the subscription active again
```

☝️ In the case of required authentication, we should persist the subscription in an invalid state. This is because we've created a `PaymentIntent` specifically against that subscription, versus a `PaymentIntent` created as a one-off payment.

I think it would be cool to give the developer two options:

1. If the subscription object requires authentication, let them grab the link to Stripe's hosted authentication page. `subscription.hosted_authentication_url`
2. The ability grab the information needed to build their own authentication screen `subscription.payment_intent_id` and `subscription.payment_intent_client_secret`

## Updating a card with Stripe

As above, I think we can keep this mostly [the same](https://github.com/jasoncharnes/pay#updating-a-customers-credit-card).

We need to provide a `SetupIntent` from Stripe:

```ruby
class PaymentMethodsController < ApplicationController
  def edit
    # under the hood, this will reach out to stripe for a off-session SetupIntent
    @payment_intent = Pay::SetupIntent.new
  end
end
```

We need the setup intent for Stripe elements to know how to generate the proper token. Then, when we go back to the server:

```ruby
user.update_card(payment_method: "pm_card_visa")
```

## Conclusion

There's without a doubt ground I'm not covering here, but I think this is a decent start. I plan to start coding on this before the end of this week.

❤️

**Comment by excid3** [comment]
This is good stuff. I feel like the new version should just use SCA always and this can be a major release for SCA support.

Do you have a preference between manual or automatic confirmation? I was leaning towards manual because we can handle things inline in the controller like we have been doing in the past.

**Comment by jasoncharnes** [comment]
I prefer the manual way, but I feel like that makes me "old school."

**Comment by excid3** [comment]
Well, automatic requires a webhook and then like...AJAX polling to another endpoint? That makes automatic seem more complicated than manual to me.

Old school and reliable is the way to go.

**Comment by jasoncharnes** [comment]
![source](https://user-images.githubusercontent.com/3144140/63441693-717f0900-c3f7-11e9-895a-3243829cef3a.gif)

**Comment by savroff** [comment]
agree with @excid3 a new version just should use SCA by default. Overall if you don't want to have 3D secure, you will just change your Radar rules in Stripe account.

**Comment by jespr** [comment]
I agree with the SCA by default 🎉

**Comment by jean-francois-labbe** [comment]
@jasoncharnes can I help you on this issue?

**Comment by jasoncharnes** [comment]
@jean-francois-labbe I know @excid3 is working on a branch of SCA. Chris, is there any way we can help with that?

**Comment by excid3** [comment]
I forgot I never replied to this. The `sca` branch is actively under development and more or less finished. This will be released as Pay v2.0.0 when it is officially done.

**Comment by excid3** [comment]
Finally released SCA support today!
